Backend/middlewares/logActivity.js

```diff
@@ -0,0 +1,42 @@
+const { PrismaClient } = require("@prisma/client");
+const prisma = new PrismaClient();
+
+/**
+ * @param {string} activityType
+ */
+const logActivity = (activityType) => {
+  return async (req, res, next) => {
+    const userId = req.user?.id;
+
+    if (userId) {
+      try {
+        // Check if the user exists in the database
+        const userExists = await prisma.user.findUnique({
+          where: { id: userId },
+        });
+
+        if (!userExists) {
+          console.error(`User with ID ${userId} does not exist.`);
+          return next();
+        }
+
+        // Log the activity
+        await prisma.userActivity.create({
+          data: {
+            userId: userId,
+            activityType: activityType || req.method + " " + req.originalUrl,
+            timestamp: new Date(),
+          },
+        });
+      } catch (error) {
+        console.error("Error logging activity:", error);
+      }
+    } else {
```

Comment:
nit: you could move this to line 10 and check if (!userId) { console.error(...); return next(); }